# Text Classification:

## Data
<pre>
1. we have total of 20 types of documents(Text files) and total 18828 documents(text files).
2. You can download data from this <a href='https://drive.google.com/open?id=1rxD15nyeIPIAZ-J2VYPrDRZI66-TBWvM'>link</a>, in that you will get documents.rar folder. <br>If you unzip that, you will get total of 18828 documnets. document name is defined as'ClassLabel_DocumentNumberInThatLabel'. 
so from document name, you can extract the label for that document.
4. Now our problem is to classify all the documents into any one of the class.
5. Below we provided count plot of all the labels in our data. 
</pre>

In [22]:
import pandas as pd
from tqdm import tqdm
import os
folder = 'C:\\Users\\user\\Desktop\\CNN\\CNN_with_textdata\\documents'
#labels = {'pos': 1, 'neg': 0}
df = pd.DataFrame()
path = os.path.join(folder)
for file in tqdm(os.listdir (path) ):
    #label=file.rsplit('_')[0]
    label=file
    with open(os.path.join(path, file),'r') as infile:
        txt = infile.read()
        df = df.append([[txt, label]],ignore_index=True)

100%|████████████████████████████████████████████████████████████████████████████| 18828/18828 [05:07<00:00, 61.29it/s]


In [23]:
df.head()

0                      1
0  From: mathew <mathew@mantis.co.uk>\nSubject: A...  alt.atheism_49960.txt
1  From: mathew <mathew@mantis.co.uk>\nSubject: A...  alt.atheism_51060.txt
2  From: I3150101@dbstu1.rz.tu-bs.de (Benedikt Ro...  alt.atheism_51119.txt
3  From: mathew <mathew@mantis.co.uk>\nSubject: R...  alt.atheism_51120.txt
4  From: strom@Watson.Ibm.Com (Rob Strom)\nSubjec...  alt.atheism_51121.txt

In [3]:
df[1].unique()

array(['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc',
       'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware',
       'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles',
       'rec.sport.baseball', 'rec.sport.hockey', 'sci.crypt',
       'sci.electronics', 'sci.med', 'sci.space',
       'soc.religion.christian', 'talk.politics.guns',
       'talk.politics.mideast', 'talk.politics.misc',
       'talk.religion.misc'], dtype=object)

## Assignment:

#### sample document
<pre>
<font color='blue'>
Subject: A word of advice
From: jcopelan@nyx.cs.du.edu (The One and Only)

In article < 65882@mimsy.umd.edu > mangoe@cs.umd.edu (Charley Wingate) writes:
>
>I've said 100 times that there is no "alternative" that should think you
>might have caught on by now.  And there is no "alternative", but the point
>is, "rationality" isn't an alternative either.  The problems of metaphysical
>and religious knowledge are unsolvable-- or I should say, humans cannot
>solve them.

How does that saying go: Those who say it can't be done shouldn't interrupt
those who are doing it.

Jim
--
Have you washed your brain today?
</font>
</pre>

In [5]:
import re
preprocessed_emails=[]
for i in range(len(df[0])):
    doc_preprocessed_email=re.findall('@[\w.-]+',df[0][i])
    df[0][i]=re.sub('@[\w.-]+',' ',df[0][i])
    preprocessed_email=[]
    for j in range(len(doc_preprocessed_email)):
        doc_preprocessed_email[j]=doc_preprocessed_email[j].lstrip('@')
        preprocessed_email.extend(doc_preprocessed_email[j].split('.'))
    duplicate=[]
    duplicate.extend(preprocessed_email)
    for word in duplicate:
        if len(word)<=2:
            preprocessed_email.remove(word)
        if word=='com'or word=='COM':
            preprocessed_email.remove(word)
    preprocessed_email=(' '.join(preprocessed_email))
    preprocessed_emails.append(preprocessed_email)

### Preprocessing:
<pre>
useful links: <a href='http://www.pyregex.com/'>http://www.pyregex.com/</a>

<font color='blue'><b>1.</b></font> Find all emails in the document and then get the text after the "@". and then split those texts by '.' 
after that remove the words whose length is less than or equal to 2 and also remove'com' word and then combine those words by space. 
In one doc, if we have 2 or more mails, get all.
<b>Eg:[test@dm1.d.com, test2@dm2.dm3.com]-->[dm1.d.com, dm3.dm4.com]-->[dm1,d,com,dm2,dm3,com]-->[dm1,dm2,dm3]-->"dm1 dm2 dm3" </b> 
append all those into one list/array. ( This will give length of 18828 sentences i.e one list for each of the document). 
Some sample output was shown below. 

> In the above sample document there are emails [jcopelan@nyx.cs.du.edu, 65882@mimsy.umd.edu, mangoe@cs.umd.edu]

preprocessing:
[jcopelan@nyx.cs.du.edu, 65882@mimsy.umd.edu, mangoe@cs.umd.edu] ==> [nyx cs du edu mimsy umd edu cs umd edu] ==> 
[nyx edu mimsy umd edu umd edu]

<font color='blue'><b>2.</b></font> Replace all the emails by space in the original text. 
</pre>

In [30]:
text=re.findall('Subject:.*',df[0][281])
text=re.sub('Subject:','',text[0])
text=re.sub('Re:','',text)
text=re.sub('\(.*\)','',text)
text

'  ? '

In [31]:
for i in tqdm(range(1000)):
    text=re.findall('Subject:.*',df[0][i])
    text=re.sub('Subject:','',text[0])
    text=re.sub('Re:','',text)
    text=re.sub('\(.*\)','',text)
    #text=re.sub('\[.*\]','',text)
    #text=re.sub('[!?><*/\-"]','',text)
    text=re.sub("'s",' is',text)
    text=re.sub("can't",'can not',text)
    text=re.sub("'ve",' have',text)
    text=re.sub("'m",' am',text)
    text=re.sub("'are",' are',text)
    text=re.sub("'ll",' will',text)
    text=re.sub("'on't",' o not',text)
    match = re.search(r'FAQ:', text)
    if match:
        text=re.findall('.*FAQ:',text)
        text=re.sub(' FAQ:','',text[0])
    
#can't -> can not, 's -> is, i've -> i have, i'm -> i am, you're -> you are, i'll --> i will 
    print(i,':',text)

  0%|                                                                                         | 0/1000 [00:00<?, ?it/s]

0 :  Alt.Atheism
1 :  Alt.Atheism
2 :   Gospel Dating
3 :   university violating separation of church/state?
4 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
5 :   A visit from the Jehovah is Witnesses
6 :   Political Atheists?
7 :   An Anecdote about Islam
8 :   <Political Atheists?
9 :   >>>>>>Pompous ass
10 :   >>>>>>Pompous ass
11 :   Keith Schneider - Stealth Poster?


  1%|▉                                                                              | 12/1000 [00:00<00:08, 116.13it/s]

12 :   Keith Schneider - Stealth Poster?
13 :   <Political Atheists?
14 :   <Political Atheists?
15 :   Political Atheists?
16 :   Don't more innocents die without the death penalty?
17 :   Ancient islamic rituals
18 :   <Political Atheists?
19 :   There must be a creator! 
20 :   Americans and Evolution
21 :   Speculations
22 :   <Political Atheists?
23 :   <Political Atheists?
24 :   university violating separation of church/state?
25 :   USAToday ad 
26 :   islamic authority over women
27 :   Ancient islamic rituals


  3%|██▏                                                                            | 28/1000 [00:00<00:07, 125.17it/s]

28 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
29 :   <Political Atheists?
30 :   <Political Atheists?
31 :   EnviroLeague
32 :  Idle questions for fellow atheists
33 :   Americans and Evolution
34 :   Genocide is Caused by Atheism
35 :   Dear Mr. Theist
36 :   <Political Atheists?
37 :   Radical Agnostic... NOT!
38 :  So what is Maddi?
39 :   The wrong and the right.
40 :   <Political Atheists?
41 :   Americans and Evolution
42 :   <Political Atheists?


  4%|███▍                                                                           | 43/1000 [00:00<00:07, 130.65it/s]

43 :   <Political Atheists?
44 :   Idle questions for fellow atheists
45 :   The Problem of Satan 
46 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
47 :   Gospel Dating
48 :  A word of advice
49 :   Concerning God is Morality 
50 :   islamic authority over women
51 :   islamic authority over women
52 :  Omnipotence 
53 :   Islamic marriage?
54 :   <Political Atheists?
55 :   <Political Atheists?
56 :   <Political Atheists?
57 :   >>>>>>Pompous ass
58 :   >>>>>>Pompous ass


  6%|████▋                                                                          | 59/1000 [00:00<00:06, 136.37it/s]

59 :   Concerning God is Morality 
60 :   Jews can not hide from keith@cco.
61 :   Jews can not hide from keith@cco.
62 :   Yet more Rushdie [ ISLAMIC LAW]
63 :   An Anecdote about Islam
64 :   Genocide is Caused by Atheism
65 :   *** The list of Biblical contradictions
66 :   A visit from the Jehovah is Witnesses
67 :   <Political Atheists?
68 :   <Political Atheists?
69 :   Omnipotence 
70 :   A Little Too Satanic
71 :   Ancient islamic rituals
72 :   Americans and Evolution
73 :   Jews can not hide from keith@cco.
74 :   Omnipotence 
75 :   A visit from the Jehovah is Witnesses


  8%|██████                                                                         | 76/1000 [00:00<00:06, 143.51it/s]

76 :  Islam vs the Jehovah is Witnesses
77 :   Had to share this
78 :   Genocide is Caused by Atheism
79 :   A visit from the Jehovah is Witnesses
80 :   Gospel Dating
81 :   Gospel Dating
82 :   Americans and Evolution
83 :   A visit from the Jehovah is Witnesses
84 :   Concerning God is Morality 
85 :   Concerning God is Morality 
86 :   An Anecdote about Islam
87 :   Concerning God is Morality 
88 :   Islam & Dress Code for women


  9%|███████                                                                        | 89/1000 [00:00<00:06, 131.85it/s]

89 :   islamic authority over women
90 :   Jews can not hide from keith@cco.
91 :   Don't more innocents die without the death penalty?
92 :   Islam & Dress Code for women
93 :   There must be a creator! 
94 :   <Political Atheists?
95 :   A Little Too Satanic
96 :   Islamic marriage?
97 :   Don't more innocents die without the death penalty?
98 :   A visit from the Jehovah is Witnesses
99 :   A visit from the Jehovah is Witnesses
100 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
101 :   A Little Too Satanic
102 :   Americans and Evolution
103 :   Alt.Atheism
104 :   Theists posting
105 :   Genocide is Caused by Atheism


 11%|████████▎                                                                     | 106/1000 [00:00<00:06, 140.53it/s]

106 :   Genocide is Caused by Atheism
107 :   Radical Agnostic... NOT!
108 :   Is Keith as ignorant as he seems?
109 :   Keith Schneider - Stealth Poster?
110 :   Keith Schneider - Stealth Poster?
111 :   Political Atheists?
112 :   <Political Atheists?
113 :   Political Atheists?
114 :   >>>>>>Pompous ass
115 :   A visit from the Jehovah is Witnesses
116 :   <Political Atheists?
117 :   <Political Atheists?
118 :   <Political Atheists?
119 :   atheist?
120 :   <Political Atheists?
121 :   An Anecdote about Islam


 12%|█████████▌                                                                    | 122/1000 [00:00<00:06, 144.36it/s]

122 :  A fundamental contradiction 
123 :   atheist?
124 :   The Inimitable Rushdie ( An Anecdote about Islam
125 :   An Anecdote about Islam
126 :   Biblical Rape
127 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
128 :   Americans and Evolution
129 :   Is Morality Constant 
130 :   <Political Atheists?
131 :   islamic authority over women
132 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
133 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
134 :   Biblical Rape
135 :   atheist?
136 :   Why is sex only allowed in marriage: Rationality ?


 14%|██████████▋                                                                   | 137/1000 [00:00<00:06, 130.16it/s]

137 :   Motto Mania
138 :   A visit from the Jehovah is Witnesses 
139 :  Benediktine Metaphysics
140 :   Gospel Dating
141 :  A Remarkable Admission
142 :   Gospel Dating
143 :   Yeah, Right
144 :   Radical Agnostic... NOT!
145 :   <Political Atheists?
146 :   Yeah, Right
147 :   Yeah, Right
148 :   Ancient islamic rituals
149 :   Ancient islamic rituals
150 :   Why is sex only allowed in marriage: Rationality ?
151 :   Ancient islamic rituals
152 :   Benediktine Metaphysics
153 :   islamic authority over women


 15%|████████████                                                                  | 154/1000 [00:01<00:06, 139.60it/s]

154 :   A visit from the Jehovah is Witnesses 
155 :  Evo. & Homosexuality 
156 :  Vonnegut/atheism
157 :   Yet more Rushdie [ ISLAMIC LAW]
158 :   Request for Support
159 :   Bill Conner:
160 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
161 :   [soc.motss, et al.] "Princeton axes matching funds for Boy Scouts"
162 :  Fluids vs Liquids
163 :   Gospel Dating
164 :   Is Morality Constant 
165 :   Gospel Dating
166 :   Contradictions
167 :  Ontology 
168 :   The Inimitable Rushdie


 17%|█████████████▏                                                                | 169/1000 [00:01<00:06, 131.87it/s]

169 :   A visit from the Jehovah is Witnesses
170 :   Ancient islamic rituals
171 :   Yeah, Right
172 :   Yet more Rushdie [ ISLAMIC LAW]
173 :   Gospel Dating
174 :   And Another THing:
175 :  Periodic Post of Charley Challenges, #3, with additions
176 :   <Political Atheists?
177 :  Request for Support
178 :   Gospel Dating
179 :   Americans and Evolution
180 :   Second Law 
181 :   Gospel Dating
182 :   Gospel Dating
183 :   Not the Omni!
184 :   IF ONLY HE KNEW


 18%|██████████████▍                                                               | 185/1000 [00:01<00:05, 139.06it/s]

185 :   Allah Akbar and Praise the Lord.
186 :   Dear Mr. Theist
187 :   Is Morality Constant 
188 :   Bill Conner:
189 :   Islam & Dress Code for women
190 :   Slavery 
191 :   Jews can not hide from keith@cco.
192 :  Alt.Atheism
193 :  Alt.Atheism
194 :  Alt.Atheism
195 :   Amusing atheists and agnostics
196 :   "So help you God" in court?  
197 :   Nicknames
198 :   Amusing atheists and agnostics
199 :  Christian Morality is


 20%|███████████████▌                                                              | 200/1000 [00:01<00:05, 141.64it/s]

200 :   KORESH IS GOD!
201 :   YOU WILL ALL GO TO HELL!!!
202 :   Amusing atheists and agnostics
203 :   Is Morality Constant 
204 :   Yet more Rushdie [ ISLAMIC LAW]
205 :   about the bible quiz answers
206 :   Amusing atheists and agnostics
207 :   some thoughts.
208 :   THE POPE IS JEWISH!
209 :   free moral agency and Jeff Clark
210 :   THE POPE IS JEWISH!
211 :   A silly question on x-tianity
212 :   I don't beleive in you either.
213 :   Genocide is Caused by Theism : Evidence?
214 :   free moral agency and Jeff Clark


 22%|████████████████▊                                                             | 215/1000 [00:01<00:05, 136.23it/s]

215 :   <Political Atheists?
216 :   New Member
217 :   free moral agency
218 :   some thoughts.
219 :   some thoughts.
220 :   Genocide is Caused by Atheism
221 :   some thoughts.
222 :   Slavery 
223 :   Slavery 
224 :   New Member
225 :   Genocide is Caused by Atheism
226 :   some thoughts.
227 :   Genocide is Caused by Atheism
228 :   some thoughts.


 23%|█████████████████▊                                                            | 229/1000 [00:01<00:05, 131.05it/s]

229 :   islamic genocide
230 :   I don't beleive in you either.
231 :   Genocide is Caused by Atheism
232 :  islamic authority [sic] over women
233 :   Feminism and Islam, again
234 :   After 2000 years, can we say that Christian Morality is
235 :   The Inimitable Rushdie ( An Anecdote about Islam
236 :   Yet more Rushdie [ ISLAMIC LAW]
237 :   Yet more Rushdie [ ISLAMIC LAW]
238 :   Genocide is Caused by Atheism
239 :   The nonexistance of Atheists?!
240 :   some thoughts.
241 :   Genocide is Caused by Atheism
242 :   The Inimitable Rushdie
243 :   A Little Too Satanic


 24%|███████████████████                                                           | 244/1000 [00:01<00:05, 131.33it/s]

244 :   note to Bobby M.
245 :   The Inimitable Rushdie ( An Anecdote about Islam
246 :   Slavery 
247 :   <Political Atheists?
248 :   After 2000 years, can we say that Christian Morality is
249 :   Where are they now?
250 :   some thoughts.
251 :   thoughts on christians
252 :   Where are they now?
253 :   <Political Atheists?
254 :   <Political Atheists?
255 :   <Political Atheists?
256 :   <Political Atheists?
257 :   <Political Atheists?
258 :   <Political Atheists?
259 :   After 2000 years, can we say that Christian Morality is


 26%|████████████████████▎                                                         | 260/1000 [00:01<00:05, 135.68it/s]

260 :   After 2000 years, can we say that Christian Morality is
261 :   After 2000 years, can we say that Christian Morality is
262 :   After 2000 years, can we say that Christian Morality is
263 :   islamic genocide
264 :   Death Penalty 
265 :   thoughts on christians
266 :   Bible Quiz
267 :  Why the bible?
268 :   <Political Atheists?
269 :   New Member
270 :   The Inimitable Rushdie
271 :   Yet more Rushdie [ ISLAMIC LAW]
272 :   Yet more Rushdie [ ISLAMIC LAW]
273 :   Yet more Rushdie [ ISLAMIC LAW]


 27%|█████████████████████▎                                                        | 274/1000 [00:02<00:05, 125.65it/s]

274 :   Yet more Rushdie [ ISLAMIC LAW]
275 :   The Inimitable Rushdie
276 :   free moral agency
277 :   note to Bobby M.
278 :   Morality? 
279 :   Morality? 
280 :   Morality? 
281 :   ? 
282 :   The Inimitable Rushdie
283 :   Free Moral Agency and Kent S.
284 :   free moral agency and Jeff Clark
285 :   <<Pompous ass
286 :   <<Pompous ass
287 :   Objective morality 
288 :   who are we to judge, Bobby?
289 :   some thoughts.
290 :   who are we to judge, Bobby?
291 :   Death Penalty 


 29%|██████████████████████▊                                                       | 292/1000 [00:02<00:05, 136.63it/s]

292 :   some thoughts.
293 :   <Political Atheists?
294 :   some thoughts.
295 :   Yet more Rushdie [ ISLAMIC LAW]
296 :   Morality? 
297 :   Moraltiy? 
298 :   "Cruel" 
299 :   <Political Atheists?
300 :   Morality? 
301 :   <<Pompous ass
302 :   thoughts on christians
303 :  It is all Mary is fault!
304 :   Rawlins debunks creationism
305 :   some thoughts.
306 :   After 2000 years, can we say that Christian Morality is
307 :   islamic authority over women
308 :   Objective morality 
309 :   <<Pompous ass
310 :   Moraltiy? 


 31%|████████████████████████▎                                                     | 311/1000 [00:02<00:04, 148.54it/s]

311 :   The Inimitable Rushdie
312 :   The Inimitable Rushdie
313 :   Slavery 
314 :   <<Pompous ass
315 :   <<Pompous ass
316 :   Morality? 
317 :   Objective morality 
318 :   New Member
319 :   <Political Atheists?
320 :   <<Pompous ass
321 :   Objective morality 
322 :   Slavery 
323 :   Morality? 
324 :   After 2000 years, can we say that Christian Morality is
325 :   Yet more Rushdie [ ISLAMIC LAW]
326 :   Morality? 
327 :   "Cruel" 
328 :   <<Pompous ass
329 :   thoughts on christians


 33%|█████████████████████████▋                                                    | 330/1000 [00:02<00:04, 156.50it/s]

330 :   <Political Atheists?
331 :   Objective morality 
332 :  Islam is caused by believing 
333 :   The Inimitable Rushdie ( An Anecdote about Islam
334 :   The Inimitable Rushdie ( An Anecdote about Islam
335 :   A silly question on x-tianity
336 :   The nonexistance of Atheists?!
337 :  Keith IS a relativist!
338 :   some thoughts.
339 :   Objective Values 'v' Scientific Accuracy 
340 :  Victims of various 'Good Fight is
341 :   Keith Schneider - Stealth Poster?
342 :   It is all Mary is fault!
343 :   Amusing atheists and agnostics
344 :   Origins of the bible.
345 :   Yet more Rushdie [ ISLAMIC LAW]
346 :   Christian Morality is
347 :   After 2000 years, can we say that Christian Morality is


 35%|███████████████████████████▏                                                  | 348/1000 [00:02<00:04, 162.42it/s]

348 :   Amusing atheists and agnostics
349 :   The Inimitable Rushdie
350 :   Amusing atheists and agnostics
351 :   Islam And Scientific Predictions 
352 :   Islam And Scientific Predictions 
353 :  Southern Baptist Convention & Freemasonry
354 :   Societally acceptable behavior
355 :   Origins of the bible.
356 :   Amusing atheists and agnostics
357 :   Requests
358 :   thoughts on christians
359 :   What is a shit shoveler to do? 
360 :   Christian Morality is
361 :   Amusing atheists and agnostics
362 :  Studies on Book of Mormon
363 :   Gulf War and Peace-niks
364 :   Amusing atheists and agnostics


 36%|████████████████████████████▍                                                 | 365/1000 [00:02<00:03, 163.75it/s]

365 :  He has risen!
366 :   Idle questions for fellow atheists
367 :   Societally acceptable behavior
368 :   islamic genocide
369 :   Amusing atheists and agnostics
370 :   Christian Morality is
371 :   Death Penalty / Gulf War
372 :   Theism and Fanatism 
373 :   YOU WILL ALL GO TO HELL!!!
374 :   YOU WILL ALL GO TO HELL!!!
375 :   free moral agency
376 :   What is a shit shoveler to do? 
377 :   Genocide is Caused by Theism : Evidence?
378 :   islamic genocide
379 :   Where are they now?
380 :   "Cruel" 
381 :   <Political Atheists?


 38%|█████████████████████████████▊                                                | 382/1000 [00:02<00:04, 143.27it/s]

382 :   some thoughts.
383 :   Who Says the Apostles Were Tortured?
384 :   <Political Atheists?
385 :   Where are they now?
386 :   Societally acceptable behavior
387 :   free moral agency
388 :   Societally acceptable behavior
389 :   Christian Morality is
390 :   Gulf War and Peace-niks
391 :   Christian Morality is
392 :   Societal basis for morality
393 :   Nicknames
394 :   free moral agency
395 :   Death Penalty / Gulf War 
396 :   Is Morality Constant 
397 :   "satanic" verses
398 :   Who Says the Apostles Were Tortured?
399 :   Who has read Rushdie is _The Satanic Verses_?


 40%|███████████████████████████████▏                                              | 400/1000 [00:02<00:03, 152.58it/s]

400 :   Is Morality Constant 
401 :   Origins of the bible.
402 :   Amusing atheists and agnostics
403 :   Christian Morality is
404 :   Gulf War / Selling Arms
405 :   Amusing atheists and agnostics
406 :   Studies on Book of Mormon
407 :   Amusing atheists and agnostics
408 :   Where are they now?
409 :  STRONG & weak Atheism
410 :   Faith and Dogma
411 :   Yet more Rushdie [ ISLAMIC LAW]
412 :   Yet more Rushdie [ ISLAMIC LAW]
413 :   Who Says the Apostles Were Tortured?
414 :   Yet more Rushdie [ ISLAMIC LAW]
415 :   Islam And Scientific Predictions 


 42%|████████████████████████████████▍                                             | 416/1000 [00:02<00:04, 136.54it/s]

416 :   Societally acceptable behavior
417 :   Gulf War 
418 :   Amusing atheists and anarchists
419 :   Who has read Rushdie is _The Satanic Ve
420 :   After 2000 years, can we say that Christian Morality is
421 :   He has risen!
422 :  College atheists
423 :   Amusing atheists and agnostics
424 :   What is a shit shoveler to do? 
425 :   Amusing atheists and agnostics
426 :   Death Penalty (was  Political Athei
427 :   Amusing atheists and agnostics
428 :   Societally acceptable behavior
429 :   You will all go to hell.
430 :   "So help you God" in court?
431 :   some thoughts.
432 :   Who Says the Apostles Were Tortured?
433 :   some thoughts.
434 :   some thoughts.


 44%|█████████████████████████████████▉                                            | 435/1000 [00:03<00:03, 149.02it/s]

435 :   thoughts on christians
436 :   Victims of various 'Good Fight is
437 :   After 2000 years, can we say that Christian Morality is
438 :   KORESH IS GOD!
439 :   Where are they now?
440 :   After 2000 years, can we say that Christian Morality is
441 :   The Inimitable Rushdie
442 :   After 2000 years, can we say that Christian Morality is
443 :   KORESH IS GOD!
444 :   THE POPE IS JEWISH!
445 :   Bible Quiz
446 :   YOU WILL ALL GO TO HELL!!!
447 :   Death Penalty / Gulf War 
448 :   Societally acceptable behavior
449 :   Societally acceptable behavior
450 :   Societal basis for morality
451 :   YOU WILL ALL GO TO HELL!!!
452 :   free moral agency


 45%|███████████████████████████████████▎                                          | 453/1000 [00:03<00:03, 140.72it/s]

453 :  TEST: IGNORE
454 :   Amusing atheists and anarchists
455 :   THE POPE IS JEWISH!
456 :   Faith and Dogma
457 :  Church o' Satan 
458 :   Who has read Rushdie is _The Satanic Verses_?
459 :   islamic genocide
460 :   The Inimitable Rushdie
461 :   Who has read Rushdie is _The Satanic Verses_?
462 :   The Universe and Black Holes, was  2000 years.....
463 :   Yet more Rushdie [ ISLAMIC LAW]
464 :   Yet more Rushdie [ ISLAMIC LAW]
465 :   "Cruel" 
466 :   "Cruel" 
467 :   <Political Atheists?
468 :   <Political Atheists?
469 :   A Little Too Satanic
470 :   Amusing atheists and agnostics
471 :   Faith and Dogma


 47%|████████████████████████████████████▊                                         | 472/1000 [00:03<00:03, 150.85it/s]

472 :   Christian Morality is
473 :   YOU WILL ALL GO TO HELL!!!
474 :   Genocide is Caused by Atheism
475 :   Genocide is Caused by Atheism
476 :   Genocide is Caused by Atheism
477 :   Faith and Dogma
478 :   Faith and Dogma
479 :   Faith and Dogma
480 :   islamic authority over women
481 :   islamic authority over women
482 :   some thoughts.
483 :   some thoughts.
484 :   thoughts on christians
485 :   thoughts on christians
486 :   Nicknames
487 :   islamic authority over women
488 :   Death Penalty 
489 :   Americans and Evolution
490 :   Amusing atheists and agnostics


 49%|██████████████████████████████████████▎                                       | 491/1000 [00:03<00:03, 158.45it/s]

491 :   thoughts on christians
492 :   Who has read Rushdie is _The Satanic Verses_?
493 :  Age of Reason Was: Who has read Rushdie is
494 :   Death Penalty (was  Political Athei
495 :   some thoughts.
496 :   thoughts on christians
497 :   "Cruel" 
498 :   Studies on Book of Mormon
499 :   Christian Morality is
500 :   Requests
501 :   Studies on Book of Mormon
502 :  Logic of Jesus?
503 :   islamic authority over women
504 :   "So help you God" in court?
505 :   Societal basis for morality
506 :   Genocide is Caused by Atheism
507 :   free moral agency


 51%|███████████████████████████████████████▌                                      | 508/1000 [00:03<00:03, 136.05it/s]

508 :   Yet more Rushdie [ ISLAMIC LAW]
509 :   After 2000 years, can we say that Christian Morality is
510 :   islamic authority over women
511 :   some thoughts.
512 :   Christian Morality is
513 :   thoughts on christians
514 :   islamic authority over women
515 :   Death Penalty 
516 :   Suggestion for "resources" FAQ
517 :   Amusing atheists and agnostics
518 :   Genocide is Caused by Atheism
519 :   islamic genocide
520 :   Objective Values 'v' Scientific Accuracy 
521 :   Islam And Scientific Predictions 
522 :   After 2000 years, can we say that Christian Morality is
523 :   Yet more Rushdie [ ISLAMIC LAW]
524 :   Death Penalty / Gulf War 
525 :   Gulf War 


 53%|█████████████████████████████████████████                                     | 526/1000 [00:03<00:03, 145.00it/s]

526 :   Gulf War and Peace-niks
527 :   Christian Morality is
528 :   Genocide is Caused by Atheism
529 :   Gulf War / Selling Arms
530 :   Yet more Rushdie [ ISLAMIC LAW]
531 :   Yet more Rushdie [ ISLAMIC LAW]
532 :   Consciousness part II - Kev Strikes Back!
533 :   Consciousness part II - Kev Strikes Back!
534 :   Consciousness part II - Kev Strikes Back!
535 :   Christian Morality is
536 :   Cults Vs. Religions?
537 :   Gulf War and Peace-niks
538 :   Death Penalty / Gulf War 
539 :   DAVID CORESH IS! GOD!
540 :   It is all Mary is fault!
541 :    Islam And Scientific Predictions (was


 54%|██████████████████████████████████████████▎                                   | 542/1000 [00:03<00:03, 141.90it/s]

542 :   Yet more Rushdie [ ISLAMIC LAW]
543 :   The Inimitable Rushdie ( An Anecdote about Islam
544 :   Jewish Settlers Demolish a Mosque in Gaza
545 :  Nostalgia
546 :   some thoughts.
547 :   Bible Quiz
548 :    Islam And Scientific Predictions (was
549 :  Bible contradictions
550 :   After 2000 years, can we say that Christian Morality is
551 :   A silly question on x-tianity
552 :   Islam And Scientific Predictions (was
553 :   The Bible and Abortion
554 :  YOU WILL ALL GO TO HELL!!!
555 :   Yet more Rushdie [ ISLAMIC LAW]
556 :   note to Bobby M., again
557 :   ISLAM: a clearer view
558 :  Cannanite genocide in the Bible
559 :  getting to the point!


 56%|███████████████████████████████████████████▋                                  | 560/1000 [00:03<00:02, 150.29it/s]

560 :   Smullyanism for the day.....
561 :   Christian Morality is
562 :   Christian Morality is
563 :  From soc.religion.christian
564 :   College atheists
565 :  some thoughts on Christian books...
566 :   What it means to be human? 
567 :   Christian Morality is
568 :  Another request for Darwin Fish
569 :   Another request for Darwin Fish
570 :   FAQ sheet
571 :   Gulf War 
572 :   Death Penalty / Gulf War 
573 :   Death Penalty / Gulf War 
574 :   Death Penalty / Gulf War 
575 :   Gulf War / Selling Arms
576 :   free moral agency


 58%|█████████████████████████████████████████████                                 | 577/1000 [00:03<00:02, 149.67it/s]

577 :  Discordian & SubGenius books, addresses etc.
578 :   free moral agency
579 :  Atheism survey
580 :   Americans and Evolution, now with free Ockham is Razor inside
581 :   some thoughts.
582 :   The Inimitable Rushdie ( An Anecdote about Islam
583 :   The Inimitable Rushdie ( An Anecdote about Islam
584 :   Objective morality 
585 :   Free Moral Agency and Kent S.
586 :   free moral agency
587 :   Religion As Cause  
588 :   Yet more Rushdie [ ISLAMIC LAW]
589 :   free moral agency
590 :   Societally acceptable behavior
591 :   Societal basis for morality
592 :   Faith and Dogma
593 :   Yet more Rushdie [ ISLAMIC LAW]
594 :   After 2000 years, can we say that Christian Morality is


 60%|██████████████████████████████████████████████▍                               | 595/1000 [00:04<00:02, 157.08it/s]

595 :   Amusing atheists and agnostics
596 :   Theism and Fanatism 
597 :   islamic authority over women
598 :   Death Penalty / Gulf War 
599 :   Gulf War / Selling Arms
600 :   Gulf War 
601 :   Gulf War and Peace-niks
602 :   STRONG & weak Atheism
603 :   Who has read Rushdie is _The Satanic Verses_?
604 :   Yet more Rushdie [ ISLAMIC LAW]
605 :   Yet more Rushdie [ ISLAMIC LAW]
606 :   After 2000 years, can we say that Christian Morality is
607 :   Studies on Book of Mormon
608 :  Science and theories
609 :   Age of Reason Was: Who has read Rushdie is
610 :   Science and theories
611 :   Theism and Fanatism 


 61%|███████████████████████████████████████████████▋                              | 612/1000 [00:04<00:02, 160.46it/s]

612 :   After 2000 years, can we say that Christian Morality is
613 :   free moral agency
614 :   Room for Metaphor?
615 :   The Universe and Black Holes, was  2000 years.....
616 :   Asimov stamp
617 :   Amusing atheists and agnostics
618 :   some thoughts.
619 :   Death Penalty 
620 :   YOU WILL ALL GO TO HELL!!!
621 :   YOU WILL ALL GO TO HELL!!!
622 :   some thoughts.
623 :   Christian Morality is
624 :   Genocide is Caused by Theism : Evidence?
625 :   Gulf War and Peace-niks
626 :   After 2000 years, can we say that Christian Morality is
627 :   Asimov stamp
628 :   Mottos to replace "In doG we trust"


 63%|█████████████████████████████████████████████████                             | 629/1000 [00:04<00:02, 152.82it/s]

629 :   islamic genocide
630 :   Asimov stamp
631 :  More Best of A.A
632 :   Are atoms real? 
633 :   free moral agency
634 :   sci.skeptic.religion 
635 :   The Inimitable Rushdie
636 :   Yet more Rushdie [ ISLAMIC LAW]
637 :   some thoughts.
638 :  Islam and Sufism 
639 :  Nicknames
640 :   Death Penalty / Gulf War
641 :   Genocide is Caused by Atheism
642 :   Atheism survey
643 :   Theism and Fanatism 
644 :   Christian Morality is
645 :   Alt.Atheism
646 :  Religion As Cause  


 65%|██████████████████████████████████████████████████▍                           | 647/1000 [00:04<00:02, 158.42it/s]

647 :  Books
648 :   College atheists
649 :   Yet more Rushdie [ ISLAMIC LAW]
650 :   Who has read Rushdie is _The Satanic Verses_?
651 :   The Inimitable Rushdie
652 :   YOU WILL ALL GO TO HELL!!!
653 :   The Inimitable Rushdie
654 :   An Anecdote about Islam
655 :   free moral agency
656 :   Yet more Rushdie [ ISLAMIC LAW]
657 :   KORESH IS GOD!
658 :   Islam And Scientific Predictions (was
659 :   Keith IS a relativist!
660 :   <Political Atheists?
661 :   Objective morality 
662 :   Yet more Rushdie [ ISLAMIC LAW]
663 :   "Cruel" 


 66%|███████████████████████████████████████████████████▊                          | 664/1000 [00:04<00:02, 150.43it/s]

664 :   "Cruel" 
665 :   The Inimitable Rushdie
666 :   Genocide is Caused by Theism : Evidence?
667 :   Societal basis for morality
668 :   High Times A Comin'!!!!!!!
669 :   High Times A Comin'!!!!!!!
670 :   "Cruel" 
671 :   Christian Morality is
672 :   Who has read Rushdie is _The Satanic Verses_?
673 :   Burden of Proof
674 :   Rawlins debunks creationism
675 :   Genocide is Caused by Theism : Evidence?
676 :   <Political Atheists?
677 :   Bayesian Statistics, theism and atheism
678 :   Asimov stamp
679 :   Where are they now?
680 :   Age of Reason Was: Who has read Rushdie is


 68%|█████████████████████████████████████████████████████                         | 681/1000 [00:04<00:02, 154.61it/s]

681 :   STRONG & weak Atheism
682 :  A PROBLEM WITH OMNIPOTENCE
683 :  Atheism
684 :   YOU WILL ALL GO TO HELL!!!
685 :   Requests
686 :   A Little Too Satanic
687 :   Societally acceptable behavior
688 :   Why Rushdie is writings are unappreciated
689 :   Yet more Rushdie [ ISLAMIC LAW]
690 :   Why?
691 :  Christianity & Logic 
692 :   some thoughts.
693 :   some thoughts.
694 :   27 fundamental beliefs of SDA
695 :   Cults Vs. Religions?
696 :   YOU WILL ALL GO TO HELL!!!


 70%|██████████████████████████████████████████████████████▎                       | 697/1000 [00:04<00:02, 144.02it/s]

697 :   Poisoning the well 
698 :   some thoughts.
699 :   some thoughts.
700 :   some thoughts.
701 :   It is all Mary is fault!
702 :   thoughts on christians
703 :   Room for Metaphor?
704 :   Moraltiy? 
705 :   YOU WILL ALL GO TO HELL!!!
706 :   Islam And Scientific Predictions 
707 :   "Cruel" 
708 :  Alt.Atheism
709 :  Alt.Atheism
710 :   The nonexistance of Atheists?!
711 :   An Anecdote about Islam


 71%|███████████████████████████████████████████████████████▌                      | 712/1000 [00:04<00:02, 127.85it/s]

712 :   The Inimitable Rushdie
713 :   A Little Too Satanic
714 :   After 2000 years, can we say that Christian Morality is
715 :   Genocide is Caused by Theism : Evidence?
716 :   thoughts on christians
717 :   Death Penalty 
718 :   Americans and Evolution
719 :   Christian Morality is
720 :  Hard/Soft == Strong/Weak.  KISS!
721 :   After 2000 years, can we say that Christian Morality is
722 :   Islamic Banks (was  Slavery
723 :  We don't need no stinking subjects!
724 :   After 2000 years, can we say that Christian Morality is
725 :   <Political Atheists?
726 :   After 2000 years, can we say that Christian Morality is


 73%|████████████████████████████████████████████████████████▋                     | 727/1000 [00:05<00:02, 132.81it/s]

727 :   Burden of Proof
728 :   Cults Vs. Religions?
729 :   Gulf War and Peace-niks
730 :   Are atoms real? 
731 :   History & texts 
732 :   After 2000 years, can we say that Christian Morality is
733 :   After 2000 years, can we say that Christian Morality is
734 :   Yet more Rushdie [ ISLAMIC LAW]
735 :   Hoban 
736 :   Death Penalty 
737 :   Faith and Dogma
738 :   Rawlins debunks creationism
739 :   Rawlins debunks creationism
740 :   Mottos to replace "In doG we trust"
741 :   After 2000 years, can we say that Christian Morality is
742 :   Death Penalty 
743 :   After 2000 years, can we say that Christian Morality is


 74%|██████████████████████████████████████████████████████████                    | 744/1000 [00:05<00:01, 140.90it/s]

744 :   Yet more Rushdie [ ISLAMIC LAW]
745 :   The Inimitable Rushdie
746 :   After 2000 years, can we say that Christian Morality is
747 :   YOU WILL ALL GO TO HELL!!!
748 :   FAKE GOD, HOLY LIES
749 :   free moral agency
750 :   free moral agency
751 :   free moral agency
752 :   free moral agency
753 :   free moral agency
754 :   free moral agency
755 :   Death Penalty / Gulf War 
756 :   The Qur'an and atheists 
757 :   free moral agency
758 :   free moral agency
759 :   free moral agency
760 :   The Inimitable Rushdie ( An Anecdote about Islam
761 :   Death Penalty 


 76%|███████████████████████████████████████████████████████████▍                  | 762/1000 [00:05<00:01, 149.16it/s]

762 :   "Cruel"
763 :   Christian Morality is
764 :  [UPI] "Mother files complaint over Boy Scouts"
765 :   free moral agency
766 :   After 2000 years, can we say that Christian Morality is
767 :   YOU WILL ALL GO TO HELL!!!
768 :   free moral agency
769 :   "Cruel"
770 :   Books
771 :   Cults Vs. Religions?
772 :   Nicknames
773 :   Asimov stamp
774 :   free moral agency
775 :   After 2000 years, can we say that Christian Morality is
776 :   After 2000 years, can we say that Christian Morality is
777 :   After 2000 years, can we say that Christian Morality is
778 :   What is a shit shoveler to do? 
779 :  FREE-ENERGY TECHNOLOGY
780 :   Christian Morality is


 78%|████████████████████████████████████████████████████████████▉                 | 781/1000 [00:05<00:01, 157.86it/s]

781 :   free moral agency
782 :   FREE-ENERGY TECHNOLOGY
783 :   The Qur'an and atheists 
784 :   Cults Vs. Religions?
785 :   Science and theories
786 :   A Little Too Satanic
787 :  Albert Sabin
788 :  Albert Sabin
789 :  After 2000 years, can we say that Christian Morality is
790 :  some thoughts.
791 :   "So help you God" in court?
792 :   Alleged Deathbed Conversions 
793 :   YOU WILL ALL GO TO HELL!!!
794 :   The Inimitable Rushdie ( An Anecdote about Islam
795 :   Islam And Scientific Predictions 
796 :   Wholly Babble 
797 :   Yet more Rushdie [ ISLAMIC LAW]
798 :   Alleged Deathbed Conversions 
799 :  CALL FOR PRESENTATIONS: Navy SciViz/VR Seminar


 80%|██████████████████████████████████████████████████████████████▍               | 800/1000 [00:05<00:01, 164.47it/s]

800 :  graphical representation of vector-valued functions
801 :   DMORPH
802 :    VOICE INPUT -- vendor information needed
803 :  Where can I find SIPP?
804 :   16 million vs 65 thousand colors
805 :  Translating TTTDDD to DXF or Swiv3D.
806 :  Help on RenderMan language wanted!
807 :   More gray levels out of the screen
808 :   More gray levels out of the screen
809 :   Rumours about 3DO ???
810 :  Visuallib 
811 :  Graphics Library Package
812 :   Rumours about 3DO ???
813 :   Graphics Library Package
814 :  WANTED: SIRD Alogorythmn
815 :  Any graphics packages available for AIX ?
816 :   BW hardcopy of colored window?
817 :   More gray levels out of the screen
818 :  LOOKING for CTDS !
819 :   POV previewer


 82%|███████████████████████████████████████████████████████████████▉              | 820/1000 [00:05<00:01, 172.69it/s]

820 :   Real Time Graphics??
821 :  BGI Drivers for SVGA
822 :  RE: VESA on the Speedstar 24
823 :  
824 :  HELP WANTED FOR DMORF.......!
825 :  test....
826 :   Rumours about 3DO ???
827 :  Technical Help Sought
828 :  Surface normal orientations
829 :   Graphics Library Package
830 :   Signature Image Database
831 :   Surface normal orientations
832 :   Genoa Blitz 24 hits 1600x1200x256 NI !
833 :   Surface normal orientations
834 :   Surface normal orientations
835 :   16 million vs 65 thousand colors
836 :   Real Time Graphics??
837 :  IGES Viewer for DOS/Windows


 84%|█████████████████████████████████████████████████████████████████▎            | 838/1000 [00:05<00:01, 153.87it/s]

838 :  computer graphics to vcr?
839 :  Stereoscopic imaging
840 :   Fractals? what good are they?
841 :  Complex  bio-medical images..
842 :  MPEG for MS-DOS
843 :  Help!!
844 :   Help! 256 colors display in C.
845 :   Rumours about 3DO ???
846 :   QuickTime movie available
847 :  QC/MSC code to view/save images
848 :  AutoCAD -> TIFF Can it be done????
849 :  Gamma-Law Correction
850 :  MPEG for x-windows MONO needed.
851 :   thining algorithm
852 :   TIFF: philosophical significance of 42
853 :   VGA 640x400 graphics mode
854 :   Weitek P9000 ?


 86%|██████████████████████████████████████████████████████████████████▋           | 855/1000 [00:05<00:00, 151.35it/s]

855 :   Delaunay Triangulation
856 :  Graphic Formats
857 :   Nice gif code
858 :   Rumours about 3DO ???
859 :   forms for curses
860 :   Trying to view POV files.....
861 :  HOT NEW 3D Software
862 :   HOT NEW 3D Software
863 :   VGA 640x400 graphics mode
864 :  TIN files & coutours
865 :  Apr 20 Toronto Siggraph Event
866 :   Delaunay Triangulation
867 :  ANSI/AIIM MS-53 Standard Image File Format
868 :   Trying to view POV files.....
869 :   looking for circle algorithm faster than Bresenhams
870 :  E-mail of Michael Abrash?


 87%|███████████████████████████████████████████████████████████████████▉          | 871/1000 [00:05<00:00, 140.77it/s]

871 :  Fonts in POV??
872 :  Polygon Reduction for Marching Cubes
873 :  dmpeg10.zip info: Another DOS MPEG decoder/player posted
874 :  Point within a polygon
875 :  POV reboots PC after memory upgrade
876 :  Good shareware paint and/or animation software for SGI?
877 :  Processing of stereo images
878 :   Weitek P9000 ?
879 :  Sun-os and 8bit ASCII graphics
880 :   looking for circle algorithm faster than Bresenhams
881 :  Geometry package
882 :   Point within a polygon
883 :   Distance between two Bezier curves
884 :   Point within a polygon
885 :   Point within a polygon
886 :   Two pointing devices in one COM-port?
887 :   Fractals? what good are they?
888 :   BGI Drivers for SVGA


 89%|█████████████████████████████████████████████████████████████████████▎        | 889/1000 [00:06<00:00, 149.97it/s]

889 :   TIFF: philosophical significance of 42
890 :  Interleaf to CGM
891 :   Hollywood Hits, Virtual Reality
892 :   Point within a polygon
893 :  Virtual Reality for X on the CHEAP!
894 :   Virtual Reality for X on the CHEAP!
895 :  PCX
896 :  What is a good assembly VGA programming book?
897 :   Fractals? what good are they?
898 :    VGA 640x400 graphics mode
899 :  Reaction-Diffusion techniques
900 :  24 bit Graphics cards
901 :   Fractal Generation of Clouds
902 :  Please Recommend 3D Graphics Library For Mac.
903 :   HOT NEW 3D Software
904 :  Raytracers: which is best?


 90%|██████████████████████████████████████████████████████████████████████▌       | 905/1000 [00:06<00:00, 134.10it/s]

905 :  Shareware
906 :   Please Recommend 3D Graphics Library For Mac.
907 :  thinning algorithm
908 :  Looking for AUTOCAD .DXF file parser
909 :  Delaunay Triangulation
910 :   cylinder and ray
911 :   Crimson 
912 :  BOOK OF KELLS CORRECTION!!!
913 :  JOB
914 :  Chosing optimal colors for colormap ?
915 :   SIGGRAPH
916 :  3d studio works changes!!!!
917 :   Rumours about 3DO ???
918 :   Please Recommend 3D Graphics Library For M
919 :  About the various DXF format questions
920 :   MPEG for x-windows MONO needed.
921 :  Compiled version of VOGL-library for PC?
922 :   Virtual Reality for X on the CHEAP!
923 :  Micro World Data Bank II ?
924 :  Looking for Dr. Bala R. Vatti is email address


 92%|████████████████████████████████████████████████████████████████████████▏     | 925/1000 [00:06<00:00, 147.23it/s]

925 :  Round VS Elliptical DOT Screens
926 :  SGI sales practices 
927 :   Please Recommend 3D Graphics Library For Mac.
928 :   Fonts in POV??
929 :   VESA on the Speedstar 24
930 :   HOT NEW 3D Software
931 :   PCX
932 :   Please Recommend 3D Graphics Library F
933 :  Need help with Mitsubishi P78U image printer
934 :  Need graph display/edit
935 :   Rumours about 3DO ???
936 :  XGA-2 info?
937 :   TIFF: philosophical significance of 42
938 :   SGI sales practices 
939 :  Rendering Software for Multi-processor Computer S
940 :  Wanted: map of the world type gifs
941 :  [POV] Having trouble bump mapping a gif to a sphere
942 :  Point of intersection of n lines
943 :  Program Included: 2 Edge Detection Algorithms!
944 :   Virtual Reality for X on the CHEAP!


 94%|█████████████████████████████████████████████████████████████████████████▋    | 945/1000 [00:06<00:00, 158.73it/s]

945 :  RE: Hot new 3D software
946 :   cylinder and ray
947 :  Sphere from 4 points?
948 :  [ANNOUNCE] Ivan Sutherland to speak at Harvard
949 :  windows imagine??!!
950 :  VGA Mode 13h Routines Available
951 :   Please Recommend 3D Graphics Library For Mac.
952 :   Please Recommend 3D Graphics Library F
953 :   Sphere from 4 points?
954 :   Sphere from 4 points?
955 :  Rayshade to DXF,RIB,etc.. ?
956 :  Siggraph 1987 Course Notes
957 :  Cobra 2.0 1-b-1 Video card HELP ME!!!!
958 :  Externel processes for 3D Studio
959 :   Text Recognition software availability
960 :   Sphere from 4 points?
961 :   Rumours about 3DO ???


 96%|███████████████████████████████████████████████████████████████████████████   | 962/1000 [00:06<00:00, 147.11it/s]

962 :  AVI file format?
963 :   VGA 640x400 graphics mode
964 :  CView answers
965 :   alt.raytrace 
966 :  SHADOW Optical Raytracing Package?
967 :  character recognition
968 :  Need a good concave -> convex polygon algorithm
969 :  Automatic layout of state diagrams
970 :  How do I compensate for photographic viewpoint and distortion?
971 :   Sphere from 4 points?
972 :  polygon orientation in DXF?
973 :  Fractals? What good are they ?
974 :   A Book I found... graphics
975 :  Speedstar 24 - how to program the TrueColor mode ?
976 :  Info needed: 2D contour plotting
977 :  VESA standard VGA/SVGA programming???
978 :   Sphere from 4 points?
979 :   VESA on the Speedstar 24
980 :   Delaunay Triangulation


 98%|████████████████████████████████████████████████████████████████████████████▌ | 981/1000 [00:06<00:00, 157.60it/s]

981 :   Rumours about 3DO ???
982 :  .GIF to .BMP
983 :  Point within a polygon 
984 :   Any graphics packages available for AIX ?
985 :   VESA standard VGA/SVGA programming???
986 :   Please Recommend 3D Graphics Library For Mac.
987 :  Questions about Plane Eqn method of Hidden Surface removal
988 :  Reprints
989 :  Church related graphics
990 :  PC fastest line/circle drawing routines: HELP!
991 :   Rumours about 3DO ???
992 :  Oak Driver NEEDED 
993 :  manipulating a hexagonal grid
994 :   .GIF to .BMP
995 :  Looking for a good book for beginners
996 :  Need some Graphics Help!
997 :  What is reverse or negative video?


100%|█████████████████████████████████████████████████████████████████████████████▊| 998/1000 [00:06<00:00, 143.44it/s]

998 :   Need a good concave -> convex polygon algorithm
999 :   Please Recommend 3D Graphics Library For Mac.


100%|█████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:06<00:00, 147.27it/s]


In [25]:
df[1][281]

'alt.atheism_53175.txt'

In [160]:
df[0][234]

'From: frank  (Frank O\'Dwyer)\nSubject: Re: After 2000 years, can we say that Christian Morality is\n\nIn article <1993Apr15.125245.12872 > MANDTBACKA  (Mats Andtbacka) writes:\n|In <1qie61$fkt > frank  writes:\n|> In article <30114 > halat  (Jim Halat) writes:\n|\n|> #I\'m one of those people who does not know what the word objective means \n|> #when put next to the word morality.  I assume its an idiom and cannot\n|> #be defined by its separate terms.\n|> #\n|> #Give it a try.\n|> \n|> Objective morality is morality built from objective values.\n|\n|      "And these objective values are ... ?"\n|Please be specific, and more importantly, motivate.\n\nI\'ll take a wild guess and say Freedom is objectively valuable.  I base\nthis on the assumption that if everyone in the world were deprived utterly\nof their freedom (so that their every act was contrary to their volition),\nalmost all would want to complain.  Therefore I take it that to assert or\nbelieve that "Freedom is not very valu

<pre>
<font color='blue'><b>3.</b></font> Get subject of the text i.e. get the total lines where "Subject:" occur and remove 
the word which are before the ":" remove the newlines, tabs, punctuations, any special chars.
<b>Eg: if we have sentance like "Subject: Re: Gospel Dating @ \r\r\n" --> You have to get "Gospel Dating"</b> 
Save all this data into another list/array. 

<font color='blue'><b>4.</b></font> After you store it in the list, Replace those sentances in original text by space.

<font color='blue'><b>5.</b></font> Delete all the sentances where sentence starts with <b>"Write to:"</b> or <b>"From:"</b>.
> In the above sample document check the 2nd line, we should remove that

<font color='blue'><b>6.</b></font> Delete all the tags like "< anyword >"
> In the above sample document check the 4nd line, we should remove that "< 65882@mimsy.umd.edu >"


<font color='blue'><b>7.</b></font> Delete all the data which are present in the brackets. 
In many text data, we observed that, they maintained the explanation of sentence 
or translation of sentence to another language in brackets so remove all those.
<b>Eg: "AAIC-The course that gets you HIRED(AAIC - Der Kurs, der Sie anstellt)" --> "AAIC-The course that gets you HIRED"</b>

> In the above sample document check the 4nd line, we should remove that "(Charley Wingate)"


<font color='blue'><b>8.</b></font> Remove all the newlines('\n'), tabs('\t'), "-", "\".

<font color='blue'><b>9.</b></font> Remove all the words which ends with <b>":"</b>.
<b>Eg: "Anyword:"</b>
> In the above sample document check the 4nd line, we should remove that "writes:"


<font color='blue'><b>10.</b></font> Decontractions, replace words like below to full words. 
please check the donors choose preprocessing for this 
<b>Eg: can't -> can not, 's -> is, i've -> i have, i'm -> i am, you're -> you are, i'll --> i will </b>

<b> There is no order to do point 6 to 10. but you have to get final output correctly</b>

<font color='blue'><b>11.</b></font> Do chunking on the text you have after above preprocessing. 
Text chunking, also referred to as shallow parsing, is a task that 
follows Part-Of-Speech Tagging and that adds more structure to the sentence.
So it combines the some phrases, named entities into single word.
So after that combine all those phrases/named entities by separating <b>"_"</b>. 
And remove the phrases/named entities if that is a "Person". 
You can use <b>nltk.ne_chunk</b> to get these. 
Below we have given one example. please go through it. 

useful links: 
<a href='https://www.nltk.org/book/ch07.html'>https://www.nltk.org/book/ch07.html</a>
<a href='https://stackoverflow.com/a/31837224/4084039'>https://stackoverflow.com/a/31837224/4084039</a>
<a href='http://www.nltk.org/howto/tree.html'>http://www.nltk.org/howto/tree.html</a>
<a href='https://stackoverflow.com/a/44294377/4084039'>https://stackoverflow.com/a/44294377/4084039</a>
</pre>

In [0]:
#i am living in the New York
print("i am living in the New York -->", list(chunks))
print(" ")
print("-"*50)
print(" ")
#My name is Srikanth Varma
print("My name is Srikanth Varma -->", list(chunks1))

i am living in the New York --> [('i', 'NN'), ('am', 'VBP'), ('living', 'VBG'), ('in', 'IN'), ('the', 'DT'), Tree('GPE', [('New', 'NNP'), ('York', 'NNP')])]
 
--------------------------------------------------
 
My name is Srikanth Varma --> [('My', 'PRP$'), ('name', 'NN'), ('is', 'VBZ'), Tree('PERSON', [('Srikanth', 'NNP'), ('Varma', 'NNP')])]


<pre>We did chunking for above two lines and then We got one list where each word is mapped to a 
POS(parts of speech) and also if you see "New York" and "Srikanth Varma", 
they got combined and represented as a tree and "New York" was referred as "GPE" and "Srikanth Varma" was referred as "PERSON". 
so now you have to Combine the "New York" with <b>"_"</b> i.e "New_York"
and remove the "Srikanth Varma" from the above sentence because it is a person.</pre>

<pre>
<font color='blue'><b>13.</b></font> Replace all the digits with space i.e delete all the digits. 
> In the above sample document, the 6th line have digit 100, so we have to remove that.

<font color='blue'><b>14.</b></font> After doing above points, we observed there might be few word's like
 <b> "_word_" (i.e starting and ending with the _), "_word" (i.e starting with the _),
  "word_" (i.e ending with the _)</b> remove the <b>_</b> from these type of words. 

<font color='blue'><b>15.</b></font>  We also observed some words like <b> "OneLetter_word"- eg: d_berlin, 
"TwoLetters_word" - eg: dr_berlin </b>, in these words we remove the "OneLetter_" (d_berlin ==> berlin) and 
"TwoLetters_" (de_berlin ==> berlin). i.e remove the words 
which are length less than or equal to 2 after spliiting those words by "_". 

<font color='blue'><b>16.</b></font> Convert all the words into lower case and lowe case 
and remove the words which are greater than or equal to 15 or less than or equal to 2.

<font color='blue'><b>17.</b></font> replace all the words except "A-Za-z_" with space. 

<font color='blue'><b>18.</b></font> Now You got Preprocessed Text, email, subject. create a dataframe with those. 
Below are the columns of the df. 
</pre>

In [0]:
data.columns

Index(['text', 'class', 'preprocessed_text', 'preprocessed_subject',
       'preprocessed_emails'],
      dtype='object')


In [0]:
data.iloc[400]

text                    From: arc1@ukc.ac.uk (Tony Curtis)\r\r\r\nSubj...
class                                                         alt.atheism
preprocessed_text       said re is article if followed the quoting rig...
preprocessed_subject                                christian morality is
preprocessed_emails                                   ukc mac macalstr edu
Name: 567, dtype: object


### To get above mentioned data frame --> Try to Write Total Preprocessing steps in One Function Named Preprocess as below. 

In [0]:
def preprocess(Input_Text):
    """Do all the Preprocessing as shown above and
    return a tuple contain preprocess_email,preprocess_subject,preprocess_text for that Text_data"""
    return (list_of_preproessed_emails,subject,text)

### Code checking:

<font color='red' size=4>
After Writing preprocess function. call that functoin with the input text of 'alt.atheism_49960' doc and print the output of the preprocess function
<br>
This will help us to evaluate faster, based on the output we can suggest you if there are any changes.
</font>

### After writing Preprocess function, call the function for each of the document(18828 docs) and then create a dataframe as mentioned above.

### Training The models to Classify: 

<pre>
1. Combine "preprocessed_text", "preprocessed_subject", "preprocessed_emails" into one column. use that column to model. 

2. Now Split the data into Train and test. use 25% for test also do a stratify split. 

3. Analyze your text data and pad the sequnce if required. 
Sequnce length is not restricted, you can use anything of your choice. 
you need to give the reasoning

4. Do Tokenizer i.e convert text into numbers. please be careful while doing it. 
if you are using tf.keras "Tokenizer" API, it removes the <b>"_"</b>, but we need that.

5. code the model's ( Model-1, Model-2 ) as discussed below 
and try to optimize that models.  

6. For every model use predefined Glove vectors. 
<b>Don't train any word vectors while Training the model.</b>

7. Use "categorical_crossentropy" as Loss. 

8. Use <b>Accuracy and Micro Avgeraged F1 score</b> as your as Key metrics to evaluate your model. 

9.  Use Tensorboard to plot the loss and Metrics based on the epoches.

10. Please save your best model weights in to <b>'best_model_L.h5' ( L = 1 or 2 )</b>. 

11. You are free to choose any Activation function, learning rate, optimizer.
But have to use the same architecture which we are giving below.

12. You can add some layer to our architecture but you <b>deletion</b> of layer is not acceptable.

13. Try to use <b>Early Stopping</b> technique or any of the callback techniques that you did in the previous assignments.

14. For Every model save your model to image ( Plot the model) with shapes 
and inlcude those images in the notebook markdown cell, 
upload those imgages to Classroom. You can use "plot_model" 
please refer <a href='https://www.tensorflow.org/api_docs/python/tf/keras/utils/plot_model'>this</a> if you don't know how to plot the model with shapes. 

</pre>

### Model-1: Using 1D convolutions with word embeddings

<pre>
<b>Encoding of the Text </b> --> For a given text data create a Matrix with Embedding layer as shown Below. 
In the example we have considered d = 5, but in this assignment we will get d = dimension of Word vectors we are using.
 i.e if we have maximum of 350 words in a sentence and embedding of 300 dim word vector, 
 we result in 350*300 dimensional matrix for each sentance as output after embedding layer
<img src='https://i.imgur.com/kiVQuk1.png'>
Ref: https://i.imgur.com/kiVQuk1.png

<b>Reference:</b>
<a href='https://stackoverflow.com/a/43399308/4084039'>https://stackoverflow.com/a/43399308/4084039</a>
<a href='https://missinglink.ai/guides/keras/keras-conv1d-working-1d-convolutional-neural-networks-keras/'>https://missinglink.ai/guides/keras/keras-conv1d-working-1d-convolutional-neural-networks-keras/</a>

<b><a href='https://stats.stackexchange.com/questions/270546/how-does-keras-embedding-layer-work'>How EMBEDDING LAYER WORKS </a></b>

</pre>

### Go through this blog, if you have any doubt on using predefined Embedding values in Embedding layer - https://machinelearningmastery.com/use-word-embedding-layers-deep-learning-keras/

<img src='https://i.imgur.com/fv1GvFJ.png'>
ref: 'https://i.imgur.com/fv1GvFJ.png'

<pre>
1. all are Conv1D layers with any number of filter and filter sizes, there is no restriction on this.

2. use concatenate layer is to concatenate all the filters/channels. 

3. You can use any pool size and stride for maxpooling layer.

4. Don't use more than 16 filters in one Conv layer becuase it will increase the no of params. 
( Only recommendation if you have less computing power )

5. You can use any number of layers after the Flatten Layer.
</pre>

### Model-2 : Using 1D convolutions with character embedding

<pre>
<pre><img src="https://i.ytimg.com/vi/CNY8VjJt-iQ/maxresdefault.jpg" width="70%">
Here are the some papers based on Char-CNN
 1. Xiang Zhang, Junbo Zhao, Yann LeCun. <a href="http://arxiv.org/abs/1509.01626">Character-level Convolutional Networks for Text Classification</a>.NIPS 2015
 2. Yoon Kim, Yacine Jernite, David Sontag, Alexander M. Rush. <a href="https://arxiv.org/abs/1508.06615">Character-Aware Neural Language Models</a>. AAAI 2016
 3. Shaojie Bai, J. Zico Kolter, Vladlen Koltun. <a href="https://arxiv.org/pdf/1803.01271.pdf">An Empirical Evaluation of Generic Convolutional and Recurrent Networks for Sequence Modeling</a>
</pre>

<img src='https://i.imgur.com/EuuoJtr.png'>